src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/course/domain/CourseService.kt

```diff
@@ -25,22 +25,57 @@ class CourseService(
     .courseOffering(offeringId)
     ?.takeIf { !it.withdrawn }
 
-  fun replaceAllCourses(courseData: List<NewCourse>) {
-    courseRepository.clear()
-    courseRepository.saveAudiences(courseData.flatMap { audienceStrings(it.audience) }.map(::Audience).toSet())
-
+  fun updateCourses(courseData: List<CourseUpdate>) {
+    updateAudiences(courseData)
     val allAudiences: Map<String, Audience> = courseRepository.allAudiences().associateBy { it.value }
-
-    courseData.map {
-      CourseEntity(
-        name = it.name,
-        identifier = it.identifier,
-        description = it.description,
-        alternateName = it.alternateName,
-        audiences = audienceStrings(it.audience).mapNotNull { audienceName -> allAudiences[audienceName] }.toMutableSet(),
-        referable = it.referable,
-      )
+    val coursesByIdentifier = courseRepository.allCourses().associateBy(CourseEntity::identifier)
+    val courseDataByIdentifier = courseData.associateBy(CourseUpdate::identifier)
+
+    val toAdd = courseDataByIdentifier.keys - coursesByIdentifier.keys
```

Comment:
as mentioned in my review on your other PR, it'd be good to be a bit more explicit about what these variables are holding in the variable name

Reply:
I agree.  I'd like to do this in a follow-up refactoring PR. Its simpler.